### ISSUE TITLE
Handling of pictures without detected face

### ISSUE THREAD

**Issue by marekkufel** [serengil/deepface#100]
Hi,

I want to test different models on my own test dataset which consists of nearly 20,000 pairs of pictures. Your doc states: ``` You should pass face pairs as array instead of calling verify function in a for loop for the best practice. This will speed the function up dramatically and reduce the allocated memory.``` So I created a list of lists with picture paths. Then I tried
```
results_positive_cosine = DeepFace.verify(result_true, detector_backend = 'ssd', distance_metric = 'cosine')
```
and got error
```
ValueError: Face could not be detected. Please confirm that the picture is a face photo or consider to set enforce_detection param to False.
``` 
so I set the enforce_detection parameter to False:
```
results_positive_cosine = DeepFace.verify(result_true, detector_backend = 'ssd', enforce_detection = False, distance_metric = 'cosine')
```
but this time I got en error:
```
error                                     Traceback (most recent call last)
<ipython-input-10-0c974523f462> in <module>
----> 1 results_positive_cosine = DeepFace.verify(result_true, detector_backend = 'ssd', enforce_detection = False, distance_metric = 'cosine')

~/anaconda3/envs/deepface/lib/python3.7/site-packages/deepface/DeepFace.py in verify(img1_path, img2_path, model_name, distance_metric, model, enforce_detection, detector_backend)
    275 
    276                         img1 = functions.preprocess_face(img=img1_path, target_size=(input_shape_y, input_shape_x), enforce_detection = enforce_detection, detector_backend = detector_backend)
--> 277                         img2 = functions.preprocess_face(img=img2_path, target_size=(input_shape_y, input_shape_x), enforce_detection = enforce_detection, detector_backend = detector_backend)
    278 
    279                         #----------------------

~/anaconda3/envs/deepface/lib/python3.7/site-packages/deepface/commons/functions.py in preprocess_face(img, target_size, grayscale, enforce_detection, detector_backend)
    457 
    458         img = detect_face(img = img, detector_backend = detector_backend, grayscale = grayscale, enforce_detection = enforce_detection)
--> 459         img = align_face(img = img, detector_backend = detector_backend)
    460 
    461         #--------------------------

~/anaconda3/envs/deepface/lib/python3.7/site-packages/deepface/commons/functions.py in align_face(img, detector_backend)
    357                 eye_detector = cv2.CascadeClassifier(eye_detector_path)
    358 
--> 359                 detected_face_gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY) #eye detector expects gray scale image
    360 
    361                 eyes = eye_detector.detectMultiScale(detected_face_gray)

error: OpenCV(4.4.0) /tmp/pip-req-build-njn2fp78/opencv/modules/imgproc/src/color.cpp:182: error: (-215:Assertion failed) !_src.empty() in function 'cvtColor'
```

The same problem occured with mtcnn detector. When I run
```
results_positive_cosine_mtcnn = DeepFace.verify(result_true, detector_backend = 'mtcnn', enforce_detection = False, distance_metric = 'cosine')
```
I got en error:
```
error                                     Traceback (most recent call last)
<ipython-input-11-228198f79070> in <module>
----> 1 results_positive_cosine_mtcnn = DeepFace.verify(result_true, detector_backend = 'mtcnn', enforce_detection = False, distance_metric = 'cosine')
      2 with open('results_positive_cosine_mtcnn.json', 'w') as f:
      3     json.dump(results_positive_cosine_mtcnn, f)

~/anaconda3/envs/deepface/lib/python3.7/site-packages/deepface/DeepFace.py in verify(img1_path, img2_path, model_name, distance_metric, model, enforce_detection, detector_backend)
    273 
    274                         img1 = functions.preprocess_face(img=img1_path, target_size=(input_shape_y, input_shape_x), enforce_detection = enforce_detection, detector_backend = detector_backend)
--> 275                         img2 = functions.preprocess_face(img=img2_path, target_size=(input_shape_y, input_shape_x), enforce_detection = enforce_detection, detector_backend = detector_backend)
    276 
    277                         #----------------------

~/anaconda3/envs/deepface/lib/python3.7/site-packages/deepface/commons/functions.py in preprocess_face(img, target_size, grayscale, enforce_detection, detector_backend)
    465                 img = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
    466 
--> 467         img = cv2.resize(img, target_size)
    468         img_pixels = image.img_to_array(img)
    469         img_pixels = np.expand_dims(img_pixels, axis = 0)

error: OpenCV(4.4.0) /tmp/pip-req-build-njn2fp78/opencv/modules/imgproc/src/resize.cpp:3929: error: (-215:Assertion failed) !ssize.empty() in function 'resize'
```

I don't know how to handle. My desired behaviour would be that if the face detector doesn't detect face then this pair of pictures is simply skipped but the rest is calculated. I can achieve this using for loop with if statement but it's against your recommendation.

**Comment by serengil** [comment]
Firstly, thank you for this deeply explained exception messages. Could you send the input image you are using?

**Comment by marekkufel** [comment]
Unfortunately I cannot send you the image. My dataset consists of real people who are not celebrities and I'm allowed to use it internally for my research but cannot share.

**Comment by serengil** [comment]
This means that I have to solve this without testing :)

**Comment by marekkufel** [comment]
In case of a picture that results in an error for mtcnn detector I noticed that my input picture shape is ```(206, 206, 3)``` and when I run your detect_face function as:
```
detect_face(img = img, detector_backend = 'mtcnn', grayscale = False, enforce_detection = False)
```
as a result I get a picture with a shape ```(0, 139, 3)```.
Then this results in an error when I try to reshape it.

**Comment by serengil** [comment]
oke, I will add try and except block tonight.

**Comment by serengil** [comment]
I just released a fix in 0.0.40 version to handle this issue. Please reinstall the library and respond me about the result.

BTW, you should set enforce_detection to False.
